feat(AIP-130): adding method AIP

toumorokoshi opened the pull request:
There's a lot of confusion around the types of methods that are possible to service teams, as well as which ones they should be using.

This is the start of an AIP to try to codify that a bit better.

The AIP starts in draft state, as I think it needs more vetting and time to determine if the terminology is correct. At the same time, there is value in adding them to the main AIPs so that they are more visible and we can contribute incremental improvements.

This is not comprehensive, and is meant to reach some agreement on the general structure of the AIP.

toumorokoshi committed:
feat(AIP-130): adding method AIP

There's a lot of confusion around the types of methods that are
possible to service teams, as well as which ones they should be using.

This is the start of an AIP to try to codify that a bit better.

The AIP starts in draft state, as I think it needs more vetting and
time to determine if the terminology is correct. At the same time,
there is value in adding them to the main AIPs so that they are
more visible and we can contribute incremental improvements.

This is not comprehensive, and is meant to reach some agreement
on the general structure of the AIP.

toumorokoshi commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+grouped up under some object (e.g. collection or resource) that the method
+operates upon.
+
+| Category Name                      | Description                                                                                                                                                                                                  | Related AIPs                                       | [IaC][]                         | CLIs        | UIs         | SDKs        |
```
I know this formatting is *really* hard to read. I'd love some suggestions on how to author a more legible markdown table.

bgrant0607 replied:
I just view the rendered file.

bgrant0607 commented on aip/general/0009.md:
```diff
@@ -108,12 +108,28 @@ A Network API exposed by a Google service. Most of these are
+
+Examples of complexities that IaC clients abstract away include:
+
+- Ordering of creation of resources
```
I would swap the order of these bullets and change this one to:
- Ordering of these imperative actions

noahdietz replied:
+1

toumorokoshi replied:
ack.

bgrant0607 commented on aip/general/0009.md:
```diff
@@ -108,12 +108,28 @@ A Network API exposed by a Google service. Most of these are
 
+[Terraform]: https://www.terraform.io/
+
 ## Changelog
```
Changelog entry?

toumorokoshi replied:
ack.

noahdietz commented on aip/general/0009.md:
```diff
@@ -108,12 +108,28 @@ A Network API exposed by a Google service. Most of these are
 
+### IaC
+
+Short for Infrastructure-as-code, Iac describes a categories of clients that
```
```suggestion
Short for Infrastructure-as-code, IaC describes a category of clients that
```

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+
+# Methods
+
+Methods are conceptually synonymous with
```
I know this is just a draft but...can you finish the sentence? The suspense is killing me!

toumorokoshi replied:
sorry! should have reviewed better. will do.

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+for any client aside from SDKs to model the operation. This category is
+preferred last due to the fact that a user cannot rely on their knowledge of
+similar APIs, as well as the issue that integration with many clients will
+likely have to be hand-writtenl.
```
```suggestion
likely have to be hand-written.
```

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+similar APIs, as well as the issue that integration with many clients will
+likely have to be hand-writtenl.
+
+As outlined in [AIP-121][],
```
```suggestion
As outlined in [AIP-121][], resources rock and every one should use them.
```

Can you finish this sentence too? :)

toumorokoshi replied:
ack.

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+| ---------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | -------------------------------------------------- | ------------------------------- | --------------- | -------------- | --------------- |
+| **Collection-oriented**            |                                                                                                                                                                                                              |                                                    |                                 |                 |                |                 |
+| Standard collection methods        | Operate on a collection of resources (List or Create).                                                                                                                                                       | [AIP-121][], [AIP-132][], [AIP-133][]              | automatable                     | automatable     | automatable    | automatable     |
+| Batch resource methods             | Fetch or mutate on multiple singular resources in a collection.                                                                                                                                              | [AIP-231][], [AIP-233][], [AIP-234][], [AIP-235][] | may be used to optimize queries | automatable     | automatable    | automatable     |
```
```suggestion
| Batch resource methods             | Fetch or mutate multiple resources in a collection by name.                                                                                                                                              | [AIP-231][], [AIP-233][], [AIP-234][], [AIP-235][] | may be used to optimize queries | automatable     | automatable    | automatable     |
```

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+| Standard collection methods        | Operate on a collection of resources (List or Create).                                                                                                                                                       | [AIP-121][], [AIP-132][], [AIP-133][]              | automatable                     | automatable     | automatable    | automatable     |
+| Batch resource methods             | Fetch or mutate on multiple singular resources in a collection.                                                                                                                                              | [AIP-231][], [AIP-233][], [AIP-234][], [AIP-235][] | may be used to optimize queries | automatable     | automatable    | automatable     |
+| Custom collection fetch methods    | Fetch information across a collection that cannot be expressed via a standard method.                                                                                                                        | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
+| Custom collection mutation methods | Perform an imperative operation referencing a collection that may mutate one or more resources within that collection in fashion that cannot be easily achieve by standard methods (e.g. state transitions). | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
```
```suggestion
| Custom collection mutation methods | Perform an imperative operation referencing a collection that may mutate one or more resources within that collection in fashion that cannot be easily achieved by standard methods (e.g. state transitions). | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
```

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+| Custom collection mutation methods | Perform an imperative operation referencing a collection that may mutate one or more resources within that collection in fashion that cannot be easily achieve by standard methods (e.g. state transitions). | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
+| **Resource-oriented**              |                                                                                                                                                                                                              |                                                    |                                 |                 |                |                 |
+| Standard resource methods          | Fetch or mutate a single resource (Get, Update, Delete).                                                                                                                                                     | [AIP-121][], [AIP-131][], [AIP-134][], [AIP-135][] | automatable                     | automatable     | automatable    | automatable     |
+| Custom resource fetch methods      | Fetch information across a single resource that cannot be expressed via a standard method.                                                                                                                   | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
```
```suggestion
| Custom resource fetch methods      | Fetch information for a single resource that cannot be expressed via a standard method.                                                                                                                   | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
```

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+| Custom resource fetch methods      | Fetch information across a single resource that cannot be expressed via a standard method.                                                                                                                   | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
+| Custom resource mutation methods   | Perform an imperative operation on a resource that may mutate it in a way a standard method cannot (e.g. state transitions).                                                                                 | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
+| **None of the above**              |                                                                                                                                                                                                              |                                                    |                                 |                 |                |                 |
+| Streaming methods                  | Methods that communicate bi-directionally.                                                                                                                                                                   |                                                    | handwritten                     | handwritten     | handwritten    | automatable     |
```
I don't think `Streaming methods` should be limited to bi-directional, because there is `client-stream` and `server-stream` as well. Folks will get hung up on that if they have proto/grpc experience.

toumorokoshi replied:
ack. Do you have a suggested way to state that? we could be general and say "methods that communicate via streaming" but that definition feels a little cyclic.

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+| **None of the above**              |                                                                                                                                                                                                              |                                                    |                                 |                 |                |                 |
+| Streaming methods                  | Methods that communicate bi-directionally.                                                                                                                                                                   |                                                    | handwritten                     | handwritten     | handwritten    | automatable     |
+
+### Preferred method of using methods
```
```suggestion
### Choosing a method category
```

noahdietz commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+
+### Preferred method of using methods
+
+Given the above categories, API authors should prefer to use the following categories, in order of preference:
```
```suggestion
While designing a method, API authors should choose from the defined categories in the following order:
```

noahdietz reviewed (commented):
Feel free to upload all of the typo/small edits in one commit instead of accepting them via GitHub.

Great start, @toumorokoshi. I like that the method types are divided by "collection" vs. "(singular) resource".

Not required, but I think those "subgroups" should still be aligned to a higher level category in the table that correlate to the categories listed in the "using methods" section. e.g. `Standard collection methods` and `Standard resource methods` both correlate to the `Standard methods` category later on. Perhaps that is a given, but being as explicit as possible near the definition is probably good too.

toumorokoshi committed:
Apply suggestions from code review

Co-authored-by: Noah Dietz <noahdietz@users.noreply.github.com>

toumorokoshi committed:
addressing feedback

- re-organizing methods by standard/custom/etc
- fixing some minor wording for IaC

noahdietz commented on aip/general/0130.md:
```diff
@@ -20,23 +21,24 @@ The following enumerates multiple categories of methods that exist, often
+| Custom resource fetch methods      | Fetch information for a single resource that cannot be expressed via a standard method.                                                                                                                       | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
+| Custom resource mutation methods   | Perform an imperative operation on a resource that may mutate it in a way a standard method cannot (e.g. state transitions).                                                                                  | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
+| **None of the above**              |                                                                                                                                                                                                               |                                                    |                                 |                 |                |                 |
+| Streaming methods                  | Methods that communicate bi-directionally.                                                                                                                                                                    |                                                    | handwritten                     | handwritten     | handwritten    | automatable     |
```
```suggestion
| Streaming methods                  | Methods that communicate via client, server, or bi-directional streams.                                                                                                                                                                   |                                                    | handwritten                     | handwritten     | handwritten    | automatable     |
```

Not that great of a suggestion, but at least it covers the bases? 😅

toumorokoshi replied:
agreed! we can always improve it: this AIP is in draft state for now.

noahdietz reviewed (commented):
Looking great @toumorokoshi

bgrant0607 commented on aip/general/0130.md:
```diff
@@ -0,0 +1,77 @@
+| Custom collection mutation methods | Perform an imperative operation referencing a collection that may mutate one or more resources within that collection in fashion that cannot be easily achieved by standard methods (e.g. state transitions). | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
+| Custom resource fetch methods      | Fetch information for a single resource that cannot be expressed via a standard method.                                                                                                                       | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
+| Custom resource mutation methods   | Perform an imperative operation on a resource that may mutate it in a way a standard method cannot (e.g. state transitions).                                                                                  | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
+| **None of the above**              |                                                                                                                                                                                                               |                                                    |                                 |                 |                |                 |
```
There are some categories of methods that are omitted:
* Aggregated list methods: automatable except for IaC, where they are not useful
* Snapshot/backup/commit and copy/restore/rollback custom methods are not really usable for IaC
* Resource rename methods are not usable for IaC
* Data-plane methods, including resourceless methods

toumorokoshi replied:
I was trying to avoid the "data-plane" term since I'm not sure that is descriptive enough as to the intent of the method, and since some things (e.g. snapshot) could be considered data-plane as well.

I can add the other ones though. I view the snapshot / backup / commit ones as more in the category of "mutation custom methods", so maybe it's worth having a whole category for that?

toumorokoshi replied:
re resourceless methods - I'm not convinced such a concept should exist. At least for GCP, everything should be done in the context of an org / folder / project. 

Although I guess categorically this refers to the more computational RPCs (e.g. speech to text)?

I could add it here, but we would need to clarify this with an AIP on that specifically I think.

bgrant0607 replied:
The management vs. data-plane distinction is important, as we've been discussing. It's mentioned in AIP 128 (https://google.aip.dev/128#guidance) but I don't think it's discussed in AIP 136. We probably need a dedicated AIP on that topic, so I don't think we need a lot of detail here, but this document seems to assume it's talking about only the management plane without saying so.

Resourceless methods are called stateless methods in AIP 136: https://google.aip.dev/136#stateless-methods. Resources often have user data attached to them, like a storage object, database table, VM memory, etc. and provide data-plane methods to write and read that data. Resourceless/stateless methods operate on data attached to other resources, such as storage objects. I believe the concept does make sense. They are pure functions, essentially. How we map the concept to an API pattern is another issue. Such a method may have an explicit parent for authorization, policy, billing, and regionalization purposes. If we want to require an explicit resource even in the case that there is no persistent state associated with the service, it would be a singleton, however we decide to represent those, so I'd suggest we tackle singleton first, then stateless methods.

toumorokoshi replied:
> The management vs. data-plane distinction is important, as we've been discussing. It's mentioned in AIP 128 (https://google.aip.dev/128#guidance) but I don't think it's discussed in AIP 136. We probably need a dedicated AIP on that topic, so I don't think we need a lot of detail here, but this document seems to assume it's talking about only the management plane without saying so.

I do agree we need to tackle management plane vs data plane, but I'm not sure about the right way to express that in this document. resource-oriented CRUDL also exists in data plane as well. 

If we do tackle the discussion of the plane, I think it makes sense to do so in this AIP.

> Resourceless methods are called stateless methods in AIP 136: https://google.aip.dev/136#stateless-methods.

Ack, I'll add a line for stateless methods. I think it's a good point that how we categorize the method and how the method is expressed via HTTP are two separate concerns. 

I don't know if I view it as a singleton though: operations on a singleton mutate data, while stateless methods do not.

toumorokoshi committed:
Update aip/general/0130.md

Co-authored-by: Noah Dietz <noahdietz@users.noreply.github.com>

toumorokoshi committed:
Merge branch 'master' into toum/add-method-categorization

toumorokoshi committed:
addressing feedback

- adding examples of custom mutation methods.

toumorokoshi committed:
Merge branch 'master' into toum/add-method-categorization

toumorokoshi committed:
addressing feedback

- adding stateless methods.

toumorokoshi committed:
Merge branch 'master' into toum/add-method-categorization

bgrant0607 commented on aip/general/0130.md:
```diff
@@ -0,0 +1,88 @@
+likely have to be hand-written.
+
+[AIP-121]: ./0121.md
+[AIP-132]: ./0132.md
```
AIP-131 is missing

bgrant0607 commented on aip/general/0130.md:
```diff
@@ -0,0 +1,88 @@
+[AIP-135]: ./0135.md
+[AIP-136]: ./0136.md
+[AIP-159]: ./0159.md
+[AIP-231]: ./0231.md
```
AIP 162 is missing. It's easy to see if you view the markdown.

toumorokoshi replied:
ack, thanks!

bgrant0607 commented on aip/general/0130.md:
```diff
@@ -0,0 +1,88 @@
+| Batch resource methods             | Fetch or mutate multiple resources in a collection by name.                                                                                                                                                   | [AIP-231][], [AIP-233][], [AIP-234][], [AIP-235][] | may be used to optimize queries | automatable     | automatable    | automatable     |
+| Aggregated list methods            | Fetch or mutate multiple resources of the same type across multiple collections.                                                                                                                              | [AIP-159][]                                        | not useful nor automable        | automatable     | automatable    | automatable     |
+| **Custom Fetch Methods**           |                                                                                                                                                                                                               |                                                    |                                 |                 |                |                 |
+| Custom collection fetch methods    | Fetch information across a collection that cannot be expressed via a standard method.                                                                                                                         | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
```
Are custom collection and resource fetch methods automatable in Terraform? Is there an example of this now?

toumorokoshi replied:
I was imagining that any custom fetch method could become a Terraform data source (the request parameters would be the resource itself).

But I can check with the Terraform team. But I guess to be on the safe side I'll put this in "handwritten".

bgrant0607 commented on aip/general/0130.md:
```diff
@@ -0,0 +1,88 @@
+| **Custom Mutation Methods**        |                                                                                                                                                                                                               |                                                    |                                 |                 |                |                 |
+| Backing up a resource              | Storing a copy of a resource at a particular point in time.                                                                                                                                                   | [AIP-162][]                                        | unused or handwritten           | automatable     | automatable    | automatable     |
+| Restoring a resource               | Setting a resource to a version from a particular point in time.                                                                                                                                              | [AIP-162][]                                        | unused or handwritten           | automatable     | automatable    | automatable     |
+| Renaming a resource                | Modify the resource's name or id while preserving configuration and state.                                                                                                                                    | [AIP-136][]                                        | unused or handwritten           | automatable     | automatable    | automatable     |
```
I suggest "data" rather than "state" since "state" has a particular meaning in AIP 216.

toumorokoshi replied:
ack, agreed.

bgrant0607 reviewed (changes requested):
A few minor fixes

toumorokoshi committed:
addressing feedback

- filling in missing links.
- fixing some wording.

toumorokoshi committed:
Merge branch 'master' into toum/add-method-categorization

bgrant0607 commented on aip/general/0130.md:
```diff
@@ -0,0 +1,90 @@
+| Custom collection mutation methods | Perform an imperative operation referencing a collection that may mutate one or more resources within that collection in fashion that cannot be easily achieved by standard methods (e.g. state transitions). | [AIP-136][]                                        | unused or handwritten           | automatable     | automatable    | automatable     |
+| Custom resource mutation methods   | Perform an imperative operation on a resource that may mutate it in a way a standard method cannot (e.g. state transitions).                                                                                  | [AIP-136][]                                        | unused or handwritten           | automatable     | automatable    | automatable     |
+| **Misc Custom Methods**            |                                                                                                                                                                                                               |                                                    |                                 |                 |                |
+| Stateless Methods                  | A method that has no permanent effect on any data within the API (e.g. translating text)                                                                                                                      | [AIP-136][]                                        | unused or handwritten           | automatable     | automtable     | automatable     |
```
typo: automtable

toumorokoshi replied:
done!

toumorokoshi committed:
fix typo
